Framework/Parallel/src/IO/EventParser.cpp

```diff
@@ -0,0 +1,285 @@
+#else
+  std::vector<size_t> sizes(data.size());
+  std::transform(data.cbegin(), data.cend(), sizes.begin(),
+                 [](const std::vector<Event> &vec) { return vec.size(); });
+  MPI_Alltoall(sizes.data(), 1, MPI_UNSIGNED_LONG, rec_sizes.data(), 1,
+               MPI_UNSIGNED_LONG, MPI_COMM_WORLD);
+
+  auto total_size = std::accumulate(rec_sizes.begin(), rec_sizes.end(),
+                                    static_cast<size_t>(0));
+  result.resize(total_size);
+  size_t offset = 0;
+  std::vector<MPI_Request> recv_requests(data.size());
+  for (int rank = 0; rank < data.size(); ++rank) {
+    int tag = 0;
+    MPI_Irecv(result.data() + offset, rec_sizes[rank] * sizeof(T), MPI_CHAR,
+              rank2, tag, MPI_COMM_WORLD, &recv_requests[rank]);
+    offset += rec_sizes[rank2];
+    // TODO:
+    // 1. do work between sending and receiving (next range, or insert into
+    // workspace after first data was received?).
+  }
+
+  std::vector<MPI_Request> send_requests(data.size());
+  for (int rank = 0; rank < data.size(); ++rank) {
+    const auto &vec = data[rank];
+    int tag = 0;
+    MPI_Isend(vec.data(), vec.size() * sizeof(T), MPI_CHAR, rank, tag,
+              MPI_COMM_WORLD, &send_requests[rank2]);
+  }
+
+  MPI_Waitall(data.size(), send_requests.data(), MPI_STATUSES_IGNORE);
+  MPI_Waitall(data.size(), recv_requests.data(), MPI_STATUSES_IGNORE);
+#endif
+}
+
+/** Fills the workspace EventList with extracted events
+ * @param eventList Workspace event list to be populated
+ * @param events Events extracted from file according to mpi rank.
+ */
+template <class IndexType, class TimeZeroType, class TimeOffsetType>
+void EventParser<IndexType, TimeZeroType, TimeOffsetType>::populateEventList(
+    std::vector<std::vector<TofEvent> *> &eventList,
+    const std::vector<EventListEntry> &events) {
+  for (const auto &event : events) {
+// TODO: calculate local index
+#ifdef MPI_EXPERIMENTAL
+    auto index = m_globalToLocalSpectrumIndex[event.index];
+#else
+    auto index = event.globalIndex;
+#endif
+    eventList[index]->emplace_back(event.tofEvent);
+    // Prefetch data into L1 Cache for faster access
+    _mm_prefetch(reinterpret_cast<char *>(&eventList[index]->back() + 1),
+                 _MM_HINT_T1);
+  }
+}
+
+/** Accepts raw data from file which has been pre-treated and sorted into chunks
+ * for parsing. The parser extracts event data from the provided buffers,
+ * separates then according to MPI ranks and then appends them to the workspace
+ * event list.
+ * @param event_id_start Buffer containing detector IDs.
+ * @param event_time_offset_start Buffer containing TOD.
+ * @param range contains information on the detector bank which corresponds to
+ * the data in the buffers, the file index offset where data starts and the
+ * number of elements in the data array.
+ */
+template <class IndexType, class TimeZeroType, class TimeOffsetType>
+void EventParser<IndexType, TimeZeroType, TimeOffsetType>::startParsing(
```

Comment:
I think it would be useful to know that what is started is done in an asynchronous manner, as you need to know this for the other methods to make sense. Can you make this `startAsync`  or `executeAsync`?